### DISCUSSION TITLE
Recording timestamp of each state transition

### DISCUSSION THREAD

**Discussion by pulasthibandara** [statelyai/xstate#2313] [Q&A]
Hi, I'm using XState for a project where I have to record the timestamp of each state that was visited. Is there a clean way to record the timestamp of each state transition on the context without having to define action on each state?

**Comment by VanTanev** [comment]
`always` on the root.

**Comment by davidkpiano** [comment]
You can do this in `.onTransition()` or `.subscribe()` and store the timestamped transitions somewhere external.

If you need those in the machine, a parallel state that's responsible for doing this can be used.

**Comment by pulasthibandara** [comment]
Thanks for the suggestions. I forgot to update this thread with what I ended up doing. I opted for using the `entry` and `exit` hooks with a HOF to decorate the nodes.

```ts
export const withStateTracking =
  (path: StatePath) =>
  <T extends StateNodeConfig<Context, TS, Event>>(node: T): T => ({
    ...node,
    entry: [...toArray(node.entry), trackStateEntry(path)],
    exit: [...toArray(node.exit), trackStateExit(path)],
  })
```

This way get more control over which states I track and the `path: StatePath` variable is basically a unique Id I assign to the state node,
